### ISSUE TITLE
Lora can not using fp16

### ISSUE THREAD

**Issue by lucasjinreal** [Lightning-AI/lit-llama#294]
I change lora to fp16 the result got totally wrong, how to enable lora fp16 mode?

**Comment by carmocca** [comment]
This is a known issue. You can find more info in this comment https://github.com/Lightning-AI/lit-llama/issues/230#issuecomment-1553585617

**Comment by rasbt** [comment]
Yeah, I think that fp16 may not have enough dynamic range to train this model well. Maybe with some additional normalization it could work, or a different hyperparameter settings. If you have time to experiment and somehow find a setting that could work, I'd be interested to know 😊! 

For the moment, let me close this since it's a duplicate. But feel free to continue the discussion in #230

**Comment by lucasjinreal** [comment]
@rasbt Oh, do u mean I trained the lora with fp16? I forgot its fp32 or fp16 trained, I mean in generate, can not set fp16, if set fp16, the result were wrong, but fp32 in inference, the finetuned model is correct

**Comment by rasbt** [comment]
@lucasjinreal I was running the finetuning with float16 and it didn't work due to NaNs in the loss after a few iterations. 
I didn't try inference with fp16, but I can imagine that the performance is also not great. 

Were you able to train with fp16?

**Comment by lucasjinreal** [comment]
@rasbt I am just trained with fp32, what I am puzzeled is I can inference with fp32 but not fp16, this is out of my expection.

**Comment by rasbt** [comment]
@lucasjinreal Do you mean in terms of memory requirements? Inference would be on a single device whereas training is on 4 GPUs. So while training usually takes more RAM than inference, the weights get sharded across the 4 GPUs so this might explain it.

**Comment by lucasjinreal** [comment]
@rasbt I mean the result, using fp16 inference totally wrong. but fp32 is OK, do they has such big differ?

**Comment by rasbt** [comment]
@lucasjinreal I see. Yes, it can make a huge difference depending on the model. Since bfloat16 seems to work fine, I suspect that there are some large weight or activation values somewhere. The max-value that bfloat16 and float32 can calculate is 10^38. In contrast, float16 can only show numbers with a max value at ~65,000

**Comment by lucasjinreal** [comment]
@rasbt it seems some other frameworks trained with fp16 lora can infer with fp16 as well successfully. i.e. huggingface peft

**Comment by lucasjinreal** [comment]
@rasbt  thank u so much for the clarify! I remmber I am training with fp16-true. but why it can infer with fp32 but not fp16

**Comment by JerryDaHeLian** [comment]
https://github.com/Lightning-AI/lit-gpt/blob/96d66b4845ebe287b5dd57b45e584b38d4f660e7/lit_gpt/speed_monitor.py#L17-L57
